lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -665,14 +678,140 @@ def write(chunk)
+    #           tialErrors",
+    #         "logEntryErrors": {
+    #           "0": {
+    #             "code": 7,
+    #             "message": "User not authorized."
+    #           },
+    #           "1": {
+    #             "code": 3,
+    #             "message": "Log name contains illegal character :"
+    #           }
+    #         }
+    #       },
+    #       {
+    #         "@type": "type.googleapis.com/google.rpc.DebugInfo",
+    #         "detail": "[ORIGINAL ERROR] generic::permission_denied: User not a
+    #           uthorized. [google.rpc.error_details_ext] { message: \"User not
+    #           authorized.\" details { type_url: \"type.googleapis.com/google.l
+    #           ogging.v2.WriteLogEntriesPartialErrors\" value: \"\\n\\034\\010\
+    #           \000\\022\\030\\010\\007\\022\\024User not authorized.\\n-\\010\
+    #           \001\\022)\\010\\003\\022%Log name contains illegal character :\
+    #           " } }"
+    #       }
+    #     ]
+    #   }
+    # }
+    #
+    # In this exxample, the logEntryErrors we are extracting should be:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    def extract_log_entry_errors(error)
+      error_details = parse_json_or_nil(error.body)['error']['details']
+      partial_errors = error_details.detect do |error_detail|
+        error_detail['@type'] == PARTIAL_ERROR_KEY
+      end
+      partial_errors['logEntryErrors']
+    rescue => e
+      @log.warn 'Failed to extract log entry errors from the error details' \
+                " json: #{error.body}.", error: e
+      {}
+    end
+
+    # Given the logEntryErrors, constuct a map from errors to a list of indexes
+    # of log entries that failed for this specific error.
+    #
+    # The logEntryErrors should be in the format of:
+    # {
+    #   "0": {
+    #     "code": 7,
+    #     "message": "User not authorized."
+    #   },
+    #   "1": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   },
+    #   "2": {
+    #     "code": 3,
+    #     "message": "Log name contains illegal character :"
+    #   }
+    # }
+    #
+    # The constructed map would be:
+    # // Keys are [error_code, error_message]. Values are a list of indexes of
+    # // log entries that failed due to this error.
+    # {
+    #   [
+    #     7,
+    #     'User not authorized.'
+    #   ]: ["0"],
+    #   [
+    #     3,
+    #     'Log name contains illegal character :'
+    #   ]: ["1", "2"]
+    # }
+    def construct_error_details_map(log_entry_errors)
+      error_details_map = {}
+      log_entry_errors.each do |index, error|
+        error_key = [error['code'], error['message']]
+        if error_details_map.key?(error_key)
+          error_details_map[error_key] << index
```

Comment:
Technically, it would eventually be useful to get indexes as ints, not strings. But for now you're only using the length of the array anyway, so let's just add a TODO to convert them to ints?

Reply:
Added.